library/Freckle/App/Csv.hs

```diff
@@ -0,0 +1,269 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+-- | Stream, parse, and validate CSVs
+--
+-- A minor extension of [cassava](https://hackage.haskell.org/package/cassava).
+-- Using `MonadValidate` and `Conduit`.
+--
+module Freckle.App.Csv
+  ( csvWithValidationSink
+  , csvWithParserAndValidationSink
+   -- * Conduit Primitives
+  , runCsvConduit
+  , decodeCsv
+  -- * Header Validation
+  , ValidateHeader
+  , validateHeader
+  , hasHeader
+  , defaultValidateOrderedHeader
+  -- * Exceptions
+  , CsvException(..)
+  -- * Options
+  , defaultOptions
+  -- * Constants
+  , headerLineNumber
+  , firstRowLineNumber
+  ) where
+
+import Freckle.App.Prelude
+
+import Conduit
+import Control.Monad (foldM)
+import Control.Monad.Validate
+  (MonadValidate(..), Validate, ValidateT, refute, runValidate, runValidateT)
+import Data.Aeson (KeyValue(..), ToJSON(..), object, pairs, (.=))
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Char8 as BS8
+import qualified Data.Conduit.Combinators as Conduit
+import qualified Data.Conduit.Text as Conduit
+import Data.Csv
+  ( DefaultOrdered
+  , FromNamedRecord(..)
+  , Header
+  , Name
+  , NamedRecord
+  , Parser
+  , defaultDecodeOptions
+  , defaultOptions
+  , headerOrder
+  )
+import qualified Data.Csv.Incremental as CsvI
+import Data.Functor.Bind (Bind)
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.List.NonEmpty as NE
+import Data.Proxy (Proxy(Proxy))
+import Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+import Data.Sequence.NonEmpty (NESeq)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Data.Vector as V
+import UnliftIO.Exception (handle)
+
+-- | Treat CSV header line as 1
+--
+-- CSVs can break rows over lines, but we don't currently handle that.
+--
+headerLineNumber :: Int
+headerLineNumber = 1
+
+-- | Start counting CSV line numbers from 2
+--
+-- We require headers for all customer CSVs, so, for line-oriented
```

Comment:
"customer"

A little business logic is leaking into our domain-agnostic toolkit here. Not blocking in the slightest, but good to keep an eye one.